sound/soc/sof/sof-audio.c

```diff
@@ -350,13 +349,35 @@ int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, in
 	/* set up widgets in the list */
 	for_each_dapm_widgets(list, num_widgets, widget) {
 		struct snd_sof_widget *swidget = widget->dobj.private;
+		struct snd_sof_widget *pipe_widget;
 
 		if (!swidget)
 			continue;
 
-		ret = sof_widget_setup(sdev, swidget);
+		/*
+		 * set up the scheduler widget for the pipeline that the widget belongs to.
+		 * The use_count for the scheduler widget is incremented for every widget
+		 * in the same pipeline to ensure that it is freed only after the last widget
+		 * in the pipeline is freed.
+		 */
+		pipe_widget = swidget->pipe_widget;
+		if (!pipe_widget) {
+			dev_err(sdev->dev, "error: no pipeline widget found for %s\n",
+				swidget->widget->name);
+			ret = -EINVAL;
+			goto widget_free;
+		}
+
+		ret = sof_widget_setup(sdev, pipe_widget);
 		if (ret < 0)
 			goto widget_free;
+
+		/* set up the widget */
+		ret = sof_widget_setup(sdev, swidget);
```

Comment:
does this mean that for a pipeline widget we call sof_widget_setup twice?

put differently, what is the defintion of swidget->pipe_widget when the swidget is *already* a pipeline :-)